keps/sig-architecture/3784-database-of-names/README.md

```diff
@@ -0,0 +1,441 @@
+3) Extend this DB to loosely-coupled projects (e.g. GKE, EKS, AKS)
+
+### Non-Goals
+
+1) Completeness.  We know that getting 100% of the worms back into the can may
+   never happen.  That doesn't mean we should not try.
+2) Enforcement.  We do not intend to inject this into code-review processes.
+   We do intend to make people AWARE of it, so they can voluntarily contribute.
+
+## Proposal
+
+The "database" should live in a new git repo (proposed name:
+"kubernetes/namesdb").  In this repo we will establish a files-and-directories
+structure which allows for easy searching and "a priori" direct indexing by
+name.  We will populate it with some of the known names from the codebase, and
+then crowd-source more.  This represents the "primary index".  We will use
+OWNERS files to delegate ownership.
+
+Given that primary index, we can programmatically produce secondary indices
+which allow easy searching by other criteria (e.g. by SIG or by API resource).
+
+We can also use this data to generate docs, if we want to make the
+"description" parts of this proposal fairly robust.
+
+### Primary index
+
+The primary index would look like:
+
+```
+github.com/kubernetes/namesdb/
+    OWNERS
+    README.md
+    by-name/
+        io/
+            k8s/
+                OWNERS
+                names.yaml
+                network/
+                    OWNERS
+                    names.yaml
+                storage/
+                    OWNERS
+                    names.yaml
+                <more...>
+            kubernetes/
+                OWNERS
+                names.yaml
+```
+
+### Secondary indices
+
+Secondary indices could be like:
+
+```
+github.com/kubernetes/namesdb/
+    by-sig/
+        network/
+            names.yaml
+        storage/
+            names.yaml
+    by-context/
+        annotation/
+            names.yaml
+        app-protocol/
+            names.yaml
+        condition/
+            names.yaml
+        label/
+            names.yaml
+        taint/
+            names.yaml
+    by-resource/
+        names.yaml
+        pod/
+            names.yaml
+        deployment/
+            names.yaml
+        endpoints/
+            names.yaml
+        service/
+            names.yaml
+```
+
+These secondary indices can be created by a simple program which we validate
+has been run before merging changes to the primary index.
+
+### Data
+
+At the "leaf" directory of each name is a YAML file which follows a defined
+schema.  The schema will certainly evolve over time, but we can start with the
+following:
+
+```go
+// NameList is the type represented in each names.yaml file.
+type NameList []Name
+
+// Name describes a single reserved name.
+type Name struct {
+    // Key is the full name being claimed, including the prefix.
+    // Required
+    Key string
+
+    // SIG specifies which SIG owns this name.  Multiple SIGs might use a name,
+    // but only one can own it.
+    // Required
+    SIG SIG
```

Comment:
is this not redundant? in your example you put this file inside a folder with the sig name

Reply:
I wrote the record so the same type could be used under multiple indices.  But there are also top-level names and type-centric names that are not clear which SIG owns.

Who owns `k8s.io/hostname` or `k8s.io/metadata.name` ?

Reply:
yeah